### ISSUE TITLE
EMQX exhook Proto order of hook execution is not always correct

### ISSUE THREAD

**Issue by RiccardoScuderi-EH** [emqx/emqx#10531]
### What happened?

I was testing the gRPC server with java and I wanted to test the "onClientAuthorize" hook when a client publishes or subscribes to the EMQX broker.
I used your project example to test that: https://github.com/emqx/emqx-extension-examples/tree/master/exhook-svr-java

Furthermore the order of the hook execution is not correct.
Sometimes the onClientAuthorize hook gets called first before the hook messagePublish or onClientSubscribe.
In the documentation is written that onClientSubscribe or messagePublish get called first and then the onClientAuthorize.
With this wrong order I can't cache the user-properties that is used for the onClientAuthorize.
The wrong execution maybe happens when I use a static method and static a variable inside onClientSubscribe or messagePublish.
I don't know if this bug comes from your java project or from your EMQX broker project.
Please take a look at that.
 

### What did you expect to happen?

The execution of onClientSubscribe or messagePublish should behave correctly. That is to say that onClientSubscribe or messsagePublish should execute before onClientAuthorize.

### How can we reproduce it (as minimally and precisely as possible)?

-register only those hooks in the main file:
             HookSpec[] specs = {HookSpec.newBuilder().setName("client.authenticate").build(),
                                HookSpec.newBuilder().setName("client.subscribe").build(),
                                HookSpec.newBuilder().setName("client.authorize").build(),
                                HookSpec.newBuilder().setName("message.publish").build()};

-Create a static HashMap
-Create a static method that puts values to the static hash map
-Cache some values of onClientSubscribe or messagePublish hook in the hash map using the static method

now use  a MQTT client and publish or subscribe to the broker and observe the order of the execution

### EMQX version

<details>
sysdescr  : EMQX
version   : 5.0.23-g15fe445c
</details>

### OS version

<details>

macOS Monterey version 12.0.1
Chip: Apple M1 Max
</details>

### Log files

<details>

</details>

**Comment by qzhuyan** [comment]
If you refer to this doc: https://www.emqx.io/docs/en/v5.0/advanced/hooks.html#hookpoint, the doc is correct. 
EMQX as broker should authorize first to see if the MQTT packet is legitimate (ACL controlled) before starting to handle the pub/sub.

For inconsistent order of hook invoking, without debug log, my best guess is the grpc request parallelism caused by a number of workers in the pool, assuming all the pub/sub are authorized.

You could try to change the pool size to 1 of the exhook server to see if you have consistent behaviors.

**Comment by RiccardoScuderi-EH** [comment]
You are right. Im referring to this doc  https://www.emqx.io/docs/en/v5.0/advanced/hooks.html#hookpoint.
I have tried to reduce the pool size to 1 but it still has this problem. 
I even made a new project of the java gRPC example that you provided:  https://github.com/emqx/emqx-extension-examples/tree/master/exhook-svr-java but still the Authorisation gets called first before the onClientSubscribe or messagePublish.

Maybe I can you this live if you have time

**Comment by qzhuyan** [comment]
right, onClientAuthorize is called after onClientSubscribe but messageOnPublish is called after onClientAuthorize.

Looks like exhooks use different grpc channel to emit hook events that the ordering is not kept. 

Could you enable debugging to confirm this?

https://www.emqx.io/docs/en/v5.0/observability/log.html

**Comment by RiccardoScuderi-EH** [comment]
Hi sorry, I was on vacation. Here is here two log file. The first is the Subscription log and the second the Publish log:
[exhook-auth-sub-log.txt](https://github.com/emqx/emqx/files/11528143/exhook-auth-sub-log.txt)
[exhook-auth-pub-log.txt](https://github.com/emqx/emqx/files/11528145/exhook-auth-pub-log.txt)

**Comment by qzhuyan** [comment]
I could confirm it is a race in grpc since we use grpc Unary call
